docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,311 @@
+misconfiguration if certain parameters that must be unique to a cluster are erroneously shared through a
+template. Since those fields cannot be automatically identified and may evolve between AKS API versions, CAPZ
+will not attempt to categorize ASO ManagedCluster fields that way like it does between the fields present and
+omitted from the `AzureClusterClassSpec` type, for example. CAPZ could document a best-effort list of known
+fields which could or should not be defined in template types and will otherwise rely on AKS to provide
+reasonable error messages for misconfigurations.
+
+Like [Option 1], this method keeps particular versions of CAPZ decoupled from particular API versions of ASO
+resources (including allowing preview versions) and opens the door for streamlined adoption of existing AKS
+clusters.
+
+#### Option 3: CAPZ resource defines an entire unstructured ASO resource inline
+
+This method is functionally equivalent to [Option 2] except that the template resource is defined inline
+within the AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ManagedClusterTemplate is the ASO ManagedCluster to be used as a template from which new
+	// ManagedClusters will be created.
+	ManagedClusterTemplate map[string]interface{} `json:"managedClusterTemplate"`
+}
+```
+
+One variant of this method could be to using `string` instead of `map[string]interface{}` for the template
+type, though that would make defining patches unwieldy (like for ClusterClass).
+
+Compared to [Option 2], this method loses schema and webhook validation that would be performed by ASO when
+creating a separate ManagedCluster to serve as a template. That validation would still be performed when CAPZ
+creates the ManagedCluster resource, but that would be some time after the AzureManagedControlPlane is created
+and error messages may not be quite as visible.
+
+#### Option 4: CAPZ resource defines an entire typed ASO resource inline
+
+This method is functionally equivalent to [Option 3] except that the template field's type is the exact same
+as an ASO ManagedCluster:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ManagedClusterSpec defines the spec of the ASO ManagedCluster managed by this AzureManagedControlPlane.
+	ManagedClusterSpec v1api20230201.ManagedCluster_Spec `json:"managedClusterTemplate"`
+}
+```
+
+This method allows CAPZ to leverage schema validation defined by ASO's CRDs upon AzureManagedControlPlane
+creation, but would still lose any further webhook validation done by ASO unless CAPZ can invoke that itself.
+
+It also has the drawback that one version of CAPZ is tied directly to a single AKS API version. The spec could
+potentially contain separate fields for each API version and enforce in webhooks that only one API version is
+being used at a time. Alternatively, users may set fields only present in a newer API version directly on the
```

Comment:
> users may set fields only present in a newer API version directly

that would be the "advanced users only, beware" backdoor but still valuable to have for break glass situations

Reply:
It also probably isn't GitOps friendly in a way that could be managed via Flux or Argo CD

Reply:
I don't think there is a concern here re: GitOps because as I understand this option - you could still "GitOps-ify" / template out (helm/kustomize) the fields set here in YAML just like anything else and not just manual edit cluster.

Reply:
@dtzar How would that work though if the ASO ManagedCluster isn't defined in YAML but generated at runtime by CAPZ based on the AzureManagedControlPlane?